File: src/Systems/Zig/patterns/command.zig
const std = @import("std");

const Command = union(enum) {
    deposit: i32,
    withdraw: i32,

    fn execute(self: Command, balance: i32) i32 {
        return switch (self) {
            .deposit => |amount| balance + amount,
            .withdraw => |amount| balance - amount,
        };
    }
};

pub fn main() void {
    const queue = [_]Command{ .{ .deposit = 50 }, .{ .withdraw = 20 } };
    var balance: i32 = 100;
    for (queue) |command| {
        balance = command.execute(balance);
    }
    std.debug.assert(balance == 130);
    std.debug.assert(queue.len == 2);
    std.debug.print("balance={d};commands={d}\n", .{ balance, queue.len });
}

File: src/Systems/Go/patterns/command.go
package main

import "fmt"

type Command interface {
	Execute(balance int) int
}

type Deposit struct{ Amount int }

func (c Deposit) Execute(balance int) int { return balance + c.Amount }

type Withdraw struct{ Amount int }

func (c Withdraw) Execute(balance int) int { return balance - c.Amount }

func main() {
	queue := []Command{Deposit{Amount: 50}, Withdraw{Amount: 20}}
	balance := 100
	for _, command := range queue {
		balance = command.Execute(balance)
	}
	if balance != 130 || len(queue) != 2 {
		panic("Command contract failed")
	}
	fmt.Printf("balance=%d;commands=%d\n", balance, len(queue))
}
